### PR TITLE
Text area test

### PR DESCRIPTION
**Pull request by Jasonjys**
1. updated README
2. create unit and integration tests
3. clean up component
4. updated the dummy app

### CODE REVIEW

File: addon/components/frost-textarea.js [hunk 1]
```diff
@@ -8,67 +15,71 @@ export default Component.extend(PropTypeMixin, {
...
    * Determine whether or not to show button for clearing out text field
    * @param {Boolean} disabled - whether or not input is disabled
+   * @param {Boolean} readonly - whether or not input is readonly
    * @param {String} value - value of text field
    * @returns {Boolean} whether or not to show button for clearing out text field
    */
-  showClear (disabled, value) {
-    return !disabled && Boolean(value)
+  showClear (disabled, readonly, value) {
+    return !disabled && !isEmpty(value) && !readonly
```
Comment (sglanzer-fire) [inline addon/components/frost-textarea.js]: We put in extra logic in the text component to handle the display/hide of the clear, might be worth duplicating here.  The base concept there was that we only wanted to show the clear when the field had focus, which would remove the need for a `readonly` check (since it can't have focus)

File: addon/styles/_frost-textarea.scss [hunk 2]
```diff
@@ -6,8 +6,8 @@ $disabled-border: 1px solid $frost-color-input-disabled-border;
 $disabled-bg: $frost-color-input-disabled-bg;
 
 .frost-textarea {
-  display: flex;
-  align-items: flex-start;
+  display: inline-block;
```
Comment (sglanzer-fire) [inline addon/styles/_frost-textarea.scss]: We need to be careful with this, flex is a block display (e.g. the next peer element will be on the next line) so if we're moving from flex we should move to `block`, any reason why `inline-block` was chosen?

File: addon/styles/_frost-textarea.scss [hunk 3]
```diff
@@ -48,17 +48,23 @@ $disabled-bg: $frost-color-input-disabled-bg;
...
-    margin: 5px 0 0 -40px;
-    width: 35px;
-    > svg {
-      width: 26px;
-      height: 26px;
-      fill: $frost-color-grey-6;
-      cursor: pointer;
-     }
-
+  &.is-clear-visible .frost-text-clear {
```
Comment (sglanzer-fire) [inline addon/styles/_frost-textarea.scss]: targets `.frost-text-clear` - should be `.frost-textarea-clear`

File: tests/integration/components/frost-textarea-test.js [hunk 4]
```diff
@@ -18,67 +17,236 @@ describeComponent(
...
     it('renders', function () {
-      this.render(hbs`{{frost-textarea}}`)
-      expect(this.$()).to.have.length(1)
+      this.render(hbs`
+        {{frost-textarea}}
+      `)
+
+      expect(
+        this.$('textarea').prop('tabindex'),
+        'tabindex is set'
```
Comment (sglanzer-fire) [inline tests/integration/components/frost-textarea-test.js:28]: Doesn't seem relevant to the test

File: tests/integration/components/frost-textarea-test.js [hunk 5]
```diff
@@ -18,67 +17,236 @@ describeComponent(
...
-        .to.be.true
-
-      expect($hook('my-textarea-input').hasClass('ember-text-area'))
-        .to.be.true
+      this.render(hbs`
+        {{frost-textarea 
+          value='Test'
+        }}
+      `)
+
```
Comment (sglanzer-fire) [inline tests/integration/components/frost-textarea-test.js:282]: Check the value prior to clicking clear as well

File: tests/unit/components/frost-textarea-test.js [hunk 6]
```diff
@@ -21,29 +22,193 @@ describeComponent(
...
+          component.set('readonly', false)
+        })
+
+        expect(
+          component.get('showClear'),
+          'showClear: true'
+        ).to.be.true
+      })
+
+      it('is set to "false" when disabled=true, value exists and readonly=false', function () {
```
Comment (sglanzer-fire) [inline tests/unit/components/frost-textarea-test.js]: The false cases would be more readable if they were packaged into a single test with an `or`

i.e. 

```
it('is set to "false" when disabled=true, value doesn't exist or readonly=true', function () {
```

**Review by sglanzer-fire — CHANGES_REQUESTED** [review]
I need confirmation on the inline-block and I wouldn't mind an extra look at the clear logic and aligning that with how the text component handles it

File: tests/integration/components/frost-textarea-test.js (hunk 4 above)
Comment (Jasonjys) [reply tests/integration/components/frost-textarea-test.js:28]: Updated the title of the test.

File: tests/integration/components/frost-textarea-test.js (hunk 5 above)
Comment (Jasonjys) [reply tests/integration/components/frost-textarea-test.js:282]: fixed

File: addon/components/frost-textarea.js (hunk 1 above)
Comment (Jasonjys) [reply addon/components/frost-textarea.js]: I tested this and while focus is not visible but it does go into readonly and that makes the clear icon show up.

File: addon/styles/_frost-textarea.scss (hunk 3 above)
Comment (Jasonjys) [reply addon/styles/_frost-textarea.scss]: fixed. This is not needed and it got removed.

File: tests/unit/components/frost-textarea-test.js (hunk 6 above)
Comment (Jasonjys) [reply tests/unit/components/frost-textarea-test.js]: Reorganized the tests but kept them separated since they are different scenarios.

File: addon/styles/_frost-textarea.scss (hunk 2 above)
Comment (Jasonjys) [reply addon/styles/_frost-textarea.scss]: Fixed positioning of the clear icon and display is now set back to flex.

**Review by Jasonjys — COMMENTED** [review]
1. fixed style sheet
2. updated test description

File: addon/templates/components/frost-textarea.hbs [hunk 7]
```diff
@@ -1,23 +1,52 @@
 {{textarea
   autofocus=autofocus
+  class='frost-textarea-input'
   cols=cols
   disabled=disabled
-  focus-out=(action 'onBlur')
   form=form
-  hook=(concat hook '-input')
+  hook=(concat receivedHook '-input')
```
Comment (juwara0) [inline addon/templates/components/frost-textarea.hbs]: you will not need this change.

File: addon/components/frost-textarea.js [hunk 8]
```diff
@@ -49,58 +65,71 @@ export default Component.extend(PropTypeMixin, {
...
-  showClear (disabled, readonly, value) {
-    return !disabled && !isEmpty(value) && !readonly
+  // // == Events ================================================================
+
+  init () {
+    this._super(...arguments)
+    this.receivedHook = this.hook
+    if (get(this, 'isHookEmbedded')) {
+      this.hook = ''
+    }
```
Comment (juwara0) [inline addon/components/frost-textarea.js]: You will not need to have this `init()` block of code. It was in `frost-text` to keep track of hooks since `frost-text` was embedded in `frost-password`.

File: addon/components/frost-textarea.js [hunk 9]
```diff
@@ -35,6 +48,9 @@ export default Component.extend(PropTypeMixin, {
   getDefaultProps () {
     return {
       autofocus: false,
+      isClearEnabled: false,
+      isClearVisible: false,
+      isHookEmbedded: false,
```
Comment (juwara0) [inline addon/components/frost-textarea.js]: You will not need `isHookEmbedded`

File: addon/components/frost-textarea.js [hunk 10]
```diff
@@ -24,6 +33,10 @@ export default Component.extend(PropTypeMixin, {
     disabled: PropTypes.bool,
     form: PropTypes.string,
     hook: PropTypes.string,
+    isClearEnabled: PropTypes.bool,
+    isClearVisible: PropTypes.bool,
+    isHookEmbedded: PropTypes.bool,
```
Comment (juwara0) [inline addon/components/frost-textarea.js]: You will not need `isHookEmbedded`

File: addon/components/frost-textarea.js [hunk 11]
```diff
@@ -24,6 +33,10 @@ export default Component.extend(PropTypeMixin, {
...
+    receivedHook: PropTypes.string,
```
Comment (juwara0) [inline addon/components/frost-textarea.js]: You will not need `receivedHook`

File: tests/unit/components/frost-textarea-test.js [hunk 12]
```diff
@@ -29,24 +30,34 @@ describeComponent(
...
-        'rows: null'
-      ).to.be.null
+        component.get('isClearVisible'),
+        'isClearVisible: false'
+      ).to.be.false
+
+      expect(
+        component.get('isHookEmbedded'),
+        'isHookEmbedded: false'
+      ).to.be.false
```
Comment (juwara0) [inline tests/unit/components/frost-textarea-test.js]: You will not need the expect() for `isHookEmbedded.

**Review by juwara0 — CHANGES_REQUESTED** [review]

File: addon/templates/components/frost-textarea.hbs (hunk 7 above)
Comment (Jasonjys) [reply addon/templates/components/frost-textarea.hbs]: fixed

File: addon/components/frost-textarea.js (hunk 10 above)
Comment (Jasonjys) [reply addon/components/frost-textarea.js]: fixed

File: addon/components/frost-textarea.js (hunk 11 above)
Comment (Jasonjys) [reply addon/components/frost-textarea.js]: fixed

File: addon/components/frost-textarea.js (hunk 9 above)
Comment (Jasonjys) [reply addon/components/frost-textarea.js]: fixed

File: tests/unit/components/frost-textarea-test.js (hunk 12 above)
Comment (Jasonjys) [reply tests/unit/components/frost-textarea-test.js]: fixed

File: addon/components/frost-textarea.js (hunk 8 above)
Comment (Jasonjys) [reply addon/components/frost-textarea.js]: fixed

File: addon/components/frost-textarea.js [hunk 13]
```diff
@@ -69,10 +66,6 @@ export default Component.extend(FrostEventsProxy, PropTypeMixin, {
 
   init () {
```
Comment (juwara0) [inline addon/components/frost-textarea.js]: Since we are no longer using `init() {}` to do some logic during initialization, the following can be removed:

```
init () {
  this._super(...arguments)
},
```

Comment (Jasonjys) [reply addon/components/frost-textarea.js]: removed

**Review by juwara0 — APPROVED** [review]
